### ISSUE TITLE
pam_unix: "try_first_pass" functionality missing

### ISSUE THREAD

**Issue by ikerexxe** [linux-pam/linux-pam#357]
Even though "try_first_pass" is mentioned in pam_unix man page, there isn't any functionality attached to it in the code. So I wonder if this is intentional and the option will be retired, or if the functionality should be there and this is a real issue. Any feedback?

In any case I'm willing to help improve it.

**Comment by t8m** [comment]
The reason is that pam_get_authtok behaves like the try_first_pass is always used. Which, IMO, makes sense. But documentation should be probably improved.

**Comment by ikerexxe** [comment]
Thank you @t8m. I'll plan for this task and I hope to work on it soon.

**Comment by OJFord** [comment]
Interesting, the behaviour I'm seeing (with v1.5.1, and thought this was going to be a report of the same) is `try_first_pass` doing what I expect from `use_first_pass`: complains the stacked password/token is wrong, and then fails without prompting.

(I noticed while testing my own module, inserted between what usually provides my password and `include system-auth` with has `pam_unix.so try_first_pass`. As a first test, I had my module `set_authtok` to an empty string, just to verify it would fail, i.e. I was indeed changing it; it did, but then I didn't get a prompt to re-enter it.)

**Comment by t8m** [comment]
Yep, the current default behavior is something in between the thing that is documented for try_first_pass and use_first_pass. It behaves as `try_first_pass` in the sense that if there is no PAM_AUTHTOK set, it will prompt for a password where the `use_first_pass` would just fail right away. On the other hand if the PAM_AUTHTOK is not a valid password, it will not prompt for a password which is more like `use_first_pass`.

**Comment by Zaba** [comment]
I've run into this issue when trying to configure `pam_p11` with fallback to password authentication. `pam_p11` saves the PIN the user entered into `PAM_AUTHTOK`, so if `pam_unix` is reached, authentication fails immediately without prompting the user for the password again, regardless of any configuration options that can be set on `pam_unix`. In what circumstances does the current behaviour make sense? Wouldn't it still be better to at least have options that control it? Or is this something that should instead be fixed in `pam_p11`, for example by making it clear `PAM_AUTHTOK`, instead of relying on `pam_unix`'s behaviour?
